// @vercel/microfrontends/dist/bin/cli.cjs
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/bin/index.ts
var import_commander = require("commander");
var import_env = require("@next/env");

// package.json
var package_default = {
  name: "@vercel/microfrontends",
  version: "2.0.1",
  private: false,
  description: "Defines configuration and utilities for microfrontends development",
  keywords: [
    "microfrontends",
    "micro-frontends",
    "micro frontends",
    "microservices",
    "Vercel",
    "Next.js",
    "React"
  ],
  homepage: "https://vercel.com/docs/microfrontends",
  repository: {
    type: "git",
    url: "https://github.com/vercel/microfrontends.git",
    directory: "packages/microfrontends"
  },
  sideEffects: false,
  type: "module",
  exports: {
    "./schema.json": "./schema/schema.json",
    "./validation": {
      import: "./dist/validation.js",
      require: "./dist/validation.cjs"
    },
    "./config": {
      import: "./dist/config.js",
      require: "./dist/config.cjs"
    },
    "./experimental/sveltekit": {
      import: "./dist/experimental/sveltekit.js",
      require: "./dist/experimental/sveltekit.cjs"
    },
    "./experimental/vite": {
      import: "./dist/experimental/vite.js",
      require: "./dist/experimental/vite.cjs"
    },
    "./overrides": {
      import: "./dist/overrides.js",
      require: "./dist/overrides.cjs"
    },
    "./microfrontends/server": {
      import: "./dist/microfrontends/server.js",
      require: "./dist/microfrontends/server.cjs"
    },
    "./microfrontends/utils": {
      import: "./dist/microfrontends/utils.js",
      require: "./dist/microfrontends/utils.cjs"
    },
    "./schema": {
      import: "./dist/schema.js",
      require: "./dist/schema.cjs"
    },
    "./next/config": {
      import: "./dist/next/config.js",
      require: "./dist/next/config.cjs"
    },
    "./next/middleware": {
      import: "./dist/next/middleware.js",
      require: "./dist/next/middleware.cjs"
    },
    "./next/testing": {
      import: "./dist/next/testing.js",
      require: "./dist/next/testing.cjs"
    },
    "./next/client": {
      import: "./dist/next/client.js",
      require: "./dist/next/client.cjs"
    },
    "./utils/mfe-port": {
      import: "./dist/utils/mfe-port.js",
      require: "./dist/utils/mfe-port.cjs"
    }
  },
  typesVersions: {
    "*": {
      validation: [
        "./dist/validation.d.ts"
      ],
      config: [
        "./dist/config.d.ts"
      ],
      "experimental/sveltekit": [
        "./dist/experimental/sveltekit.d.ts"
      ],
      "experimental/vite": [
        "./dist/experimental/vite.d.ts"
      ],
      overrides: [
        "./dist/overrides.d.ts"
      ],
      "microfrontends/server": [
        "./dist/microfrontends/server.d.ts"
      ],
      "microfrontends/utils": [
        "./dist/microfrontends/utils.d.ts"
      ],
      schema: [
        "./dist/schema.d.ts"
      ],
      "next/config": [
        "./dist/next/config.d.ts"
      ],
      "next/middleware": [
        "./dist/next/middleware.d.ts"
      ],
      "next/testing": [
        "./dist/next/testing.d.ts"
      ],
      "next/client": [
        "./dist/next/client.d.ts"
      ],
      "utils/mfe-port": [
        "./dist/utils/mfe-port.d.ts"
      ]
    }
  },
  bin: {
    microfrontends: "./cli/index.cjs"
  },
  files: [
    "dist",
    "schema",
    "CHANGELOG.md"
  ],
  scripts: {
    build: "tsup",
    postbuild: "pnpm generate:exports",
    "generate:exports": "tsx scripts/generate-exports/index.ts",
    "generate:schema": "tsx scripts/generate-json-schema.ts",
    lint: "eslint .",
    "lint-fix": "eslint . --fix",
    prepublishOnly: "pnpm build && pnpm generate:exports && pnpm generate:schema",
    proxy: "tsx src/proxy/index.ts",
    test: "cross-env TZ=UTC jest",
    typecheck: "tsc --noEmit"
  },
  dependencies: {
    "@next/env": "15.5.4",
    "@types/md5": "^2.3.5",
    ajv: "^8.17.1",
    commander: "^12.1.0",
    cookie: "1.0.2",
    "fast-glob": "^3.3.2",
    "http-proxy": "^1.18.1",
    "jsonc-parser": "^3.3.1",
    md5: "^2.3.0",
    nanoid: "^3.3.9",
    "path-to-regexp": "6.2.1",
    semver: "^7.7.2"
  },
  devDependencies: {
    "@edge-runtime/jest-environment": "^4.0.0",
    "@edge-runtime/types": "^3.0.2",
    "@sveltejs/kit": "2.17.2",
    "@testing-library/react": "^15.0.7",
    "@types/cookie": "0.5.1",
    "@types/http-proxy": "^1.17.15",
    "@types/jest": "^29.2.0",
    "@types/json-schema": "^7.0.15",
    "@types/node": "20.11.30",
    "@types/react": "18.3.1",
    "@types/react-dom": "18.3.0",
    "@types/semver": "^7.7.0",
    "eslint-config-custom": "workspace:*",
    jest: "^29.7.0",
    "jest-environment-jsdom": "29.2.2",
    next: "15.5.4",
    react: "19.0.0",
    "react-dom": "19.0.0",
    "ts-config": "workspace:*",
    "ts-json-schema-generator": "^1.1.2",
    "ts-node": "~10.9.2",
    tsup: "^6.6.2",
    tsx: "^4.6.2",
    typescript: "5.7.3",
    vite: "5.4.11",
    webpack: "5"
  },
  peerDependencies: {
    "@sveltejs/kit": ">=1",
    "@vercel/analytics": ">=1.5.0",
    "@vercel/speed-insights": ">=1.2.0",
    next: ">=13",
    react: ">=17.0.0",
    "react-dom": ">=17.0.0",
    vite: ">=5"
  },
  peerDependenciesMeta: {
    "@sveltejs/kit": {
      optional: true
    },
    "@vercel/analytics": {
      optional: true
    },
    "@vercel/speed-insights": {
      optional: true
    },
    next: {
      optional: true
    },
    react: {
      optional: true
    },
    "react-dom": {
      optional: true
    },
    vite: {
      optional: true
    }
  }
};

// src/bin/local-proxy.ts
var http = __toESM(require("http"), 1);
var https = __toESM(require("https"), 1);
var import_node_url = require("url");
var import_cookie = require("cookie");
var import_path_to_regexp3 = require("path-to-regexp");
var import_http_proxy = __toESM(require("http-proxy"), 1);

// src/config/microfrontends-config/isomorphic/index.ts
var import_jsonc_parser = require("jsonc-parser");

// src/config/errors.ts
var MicrofrontendError = class extends Error {
  constructor(message, opts) {
    super(message, { cause: opts?.cause });
    this.name = "MicrofrontendsError";
    this.source = opts?.source ?? "@vercel/microfrontends";
    this.type = opts?.type ?? "unknown";
    this.subtype = opts?.subtype;
    Error.captureStackTrace(this, MicrofrontendError);
  }
  isKnown() {
    return this.type !== "unknown";
  }
  isUnknown() {
    return !this.isKnown();
  }
  /**
   * Converts an error to a MicrofrontendsError.
   * @param original - The original error to convert.
   * @returns The converted MicrofrontendsError.
   */
  static convert(original, opts) {
    if (opts?.fileName) {
      const err = MicrofrontendError.convertFSError(original, opts.fileName);
      if (err) {
        return err;
      }
    }
    if (original.message.includes(
      "Code generation from strings disallowed for this context"
    )) {
      return new MicrofrontendError(original.message, {
        type: "config",
        subtype: "unsupported_validation_env",
        source: "ajv"
      });
    }
    return new MicrofrontendError(original.message);
  }
  static convertFSError(original, fileName) {
    if (original instanceof Error && "code" in original) {
      if (original.code === "ENOENT") {
        return new MicrofrontendError(`Could not find "${fileName}"`, {
          type: "config",
          subtype: "unable_to_read_file",
          source: "fs"
        });
      }
      if (original.code === "EACCES") {
        return new MicrofrontendError(
          `Permission denied while accessing "${fileName}"`,
          {
            type: "config",
            subtype: "invalid_permissions",
            source: "fs"
          }
        );
      }
    }
    if (original instanceof SyntaxError) {
      return new MicrofrontendError(
        `Failed to parse "${fileName}": Invalid JSON format.`,
        {
          type: "config",
          subtype: "invalid_syntax",
          source: "fs"
        }
      );
    }
    return null;
  }
  /**
   * Handles an unknown error and returns a MicrofrontendsError instance.
   * @param err - The error to handle.
   * @returns A MicrofrontendsError instance.
   */
  static handle(err, opts) {
    if (err instanceof MicrofrontendError) {
      return err;
    }
    if (err instanceof Error) {
      return MicrofrontendError.convert(err, opts);
    }
    if (typeof err === "object" && err !== null) {
      if ("message" in err && typeof err.message === "string") {
        return MicrofrontendError.convert(new Error(err.message), opts);
      }
    }
    return new MicrofrontendError("An unknown error occurred");
  }
};

// src/config/microfrontends-config/utils/get-config-from-env.ts
function getConfigStringFromEnv() {
  const config = process.env.MFE_CONFIG;
  if (!config) {
    throw new MicrofrontendError(`Missing "MFE_CONFIG" in environment.`, {
      type: "config",
      subtype: "not_found_in_env"
    });
  }
  return config;
}

// src/config/schema/utils/is-default-app.ts
function isDefaultApp(a) {
  return !("routing" in a);
}

// src/config/overrides/constants.ts
var OVERRIDES_COOKIE_PREFIX = "vercel-micro-frontends-override";
var OVERRIDES_ENV_COOKIE_PREFIX = `${OVERRIDES_COOKIE_PREFIX}:env:`;

// src/config/overrides/is-override-cookie.ts
function isOverrideCookie(cookie) {
  return Boolean(cookie.name?.startsWith(OVERRIDES_COOKIE_PREFIX));
}

// src/config/overrides/get-override-from-cookie.ts
function getOverrideFromCookie(cookie) {
  if (!isOverrideCookie(cookie) || !cookie.value)
    return;
  return {
    application: cookie.name.replace(OVERRIDES_ENV_COOKIE_PREFIX, ""),
    host: cookie.value
  };
}

// src/config/overrides/parse-overrides.ts
function parseOverrides(cookies) {
  const overridesConfig = { applications: {} };
  cookies.forEach((cookie) => {
    const override = getOverrideFromCookie(cookie);
    if (!override)
      return;
    overridesConfig.applications[override.application] = {
      environment: { host: override.host }
    };
  });
  return overridesConfig;
}

// src/config/microfrontends-config/client/index.ts
var import_path_to_regexp = require("path-to-regexp");
var regexpCache = /* @__PURE__ */ new Map();
var getRegexp = (path7) => {
  const existing = regexpCache.get(path7);
  if (existing) {
    return existing;
  }
  const regexp = (0, import_path_to_regexp.pathToRegexp)(path7);
  regexpCache.set(path7, regexp);
  return regexp;
};
var MicrofrontendConfigClient = class {
  constructor(config, opts) {
    this.pathCache = {};
    this.hasFlaggedPaths = config.hasFlaggedPaths ?? false;
    for (const app of Object.values(config.applications)) {
      if (app.routing) {
        if (app.routing.some((match) => match.flag)) {
          this.hasFlaggedPaths = true;
        }
        const newRouting = [];
        const pathsWithoutFlags = [];
        for (const group of app.routing) {
          if (group.flag) {
            if (opts?.removeFlaggedPaths) {
              continue;
            }
            if (group.group) {
              delete group.group;
            }
            newRouting.push(group);
          } else {
            pathsWithoutFlags.push(...group.paths);
          }
        }
        if (pathsWithoutFlags.length > 0) {
          newRouting.push({ paths: pathsWithoutFlags });
        }
        app.routing = newRouting;
      }
    }
    this.serialized = config;
    if (this.hasFlaggedPaths) {
      this.serialized.hasFlaggedPaths = this.hasFlaggedPaths;
    }
    this.applications = config.applications;
  }
  /**
   * Create a new `MicrofrontendConfigClient` from a JSON string.
   * Config must be passed in to remain framework agnostic
   */
  static fromEnv(config) {
    if (!config) {
      throw new Error(
        "Could not construct MicrofrontendConfigClient: configuration is empty or undefined. Did you set up your application with `withMicrofrontends`? Is the local proxy running and this application is being accessed via the proxy port? See https://vercel.com/docs/microfrontends/local-development#setting-up-microfrontends-proxy"
      );
    }
    return new MicrofrontendConfigClient(JSON.parse(config));
  }
  isEqual(other) {
    return this === other || JSON.stringify(this.applications) === JSON.stringify(other.applications);
  }
  getApplicationNameForPath(path7) {
    if (!path7.startsWith("/")) {
      throw new Error(`Path must start with a /`);
    }
    if (this.pathCache[path7]) {
      return this.pathCache[path7];
    }
    const pathname = new URL(path7, "https://example.com").pathname;
    for (const [name, application] of Object.entries(this.applications)) {
      if (application.routing) {
        for (const group of application.routing) {
          for (const childPath of group.paths) {
            const regexp = getRegexp(childPath);
            if (regexp.test(pathname)) {
              this.pathCache[path7] = name;
              return name;
            }
          }
        }
      }
    }
    const defaultApplication = Object.entries(this.applications).find(
      ([, application]) => application.default
    );
    if (!defaultApplication) {
      return null;
    }
    this.pathCache[path7] = defaultApplication[0];
    return defaultApplication[0];
  }
  serialize() {
    return this.serialized;
  }
};

// src/config/microfrontends-config/isomorphic/validation.ts
var import_path_to_regexp2 = require("path-to-regexp");
var LIST_FORMATTER = new Intl.ListFormat("en", {
  style: "long",
  type: "conjunction"
});
var VALID_ASSET_PREFIX_REGEXP = /^[a-z](?:[a-z0-9-]*[a-z0-9])?$/;
var validateConfigPaths = (applicationConfigsById) => {
  if (!applicationConfigsById) {
    return;
  }
  const pathsByApplicationId = /* @__PURE__ */ new Map();
  const errors = [];
  for (const [id, app] of Object.entries(applicationConfigsById)) {
    if (isDefaultApp(app)) {
      continue;
    }
    for (const pathMatch of app.routing) {
      for (const path7 of pathMatch.paths) {
        const maybeError = validatePathExpression(path7);
        if (maybeError) {
          errors.push(maybeError);
        } else {
          const existing = pathsByApplicationId.get(path7);
          if (existing) {
            existing.applications.push(id);
          } else {
            pathsByApplicationId.set(path7, {
              applications: [id],
              matcher: (0, import_path_to_regexp2.pathToRegexp)(path7),
              applicationId: id
            });
          }
        }
      }
    }
  }
  const entries = Array.from(pathsByApplicationId.entries());
  for (const [path7, { applications: ids, matcher, applicationId }] of entries) {
    if (ids.length > 1) {
      errors.push(
        `Duplicate path "${path7}" for applications "${ids.join(", ")}"`
      );
    }
    for (const [
      matchPath,
      { applications: matchIds, applicationId: matchApplicationId }
    ] of entries) {
      if (path7 === matchPath) {
        continue;
      }
      if (applicationId === matchApplicationId) {
        continue;
      }
      if (matcher.test(matchPath)) {
        const source = `"${path7}" of application${ids.length > 0 ? "s" : ""} ${ids.join(", ")}`;
        const destination = `"${matchPath}" of application${matchIds.length > 0 ? "s" : ""} ${matchIds.join(", ")}`;
        errors.push(
          `Overlapping path detected between ${source} and ${destination}`
        );
      }
    }
  }
  if (errors.length) {
    throw new MicrofrontendError(
      `Invalid paths: ${errors.join(", ")}. See supported paths in the documentation https://vercel.com/docs/microfrontends/path-routing#supported-path-expressions.`,
      {
        type: "config",
        subtype: "conflicting_paths"
      }
    );
  }
};
var PATH_DEFAULT_PATTERN = "[^\\/#\\?]+?";
function validatePathExpression(path7) {
  try {
    const tokens = (0, import_path_to_regexp2.parse)(path7);
    if (/(?<!\\)\{/.test(path7)) {
      return `Optional paths are not supported: ${path7}`;
    }
    if (/(?<!\\|\()\?/.test(path7)) {
      return `Optional paths are not supported: ${path7}`;
    }
    if (/\/[^/]*(?<!\\):[^/]*(?<!\\):[^/]*/.test(path7)) {
      return `Only one wildcard is allowed per path segment: ${path7}`;
    }
    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];
      if (token === void 0) {
        return `token ${i} in ${path7} is undefined, this shouldn't happen`;
      }
      if (typeof token !== "string") {
        if (!token.name) {
          return `Only named wildcards are allowed: ${path7} (hint: add ":path" to the wildcard)`;
        }
        if (token.pattern !== PATH_DEFAULT_PATTERN && // Allows (a|b|c) and ((?!a|b|c).*) regex
        // Only limited regex is supported for now, due to performance considerations
        // Allows all letters, numbers, and hyphens. Other characters must be escaped.
        !/^(?<allowed>[\w-~]+(?:\|[^:|()]+)+)$|^\(\?!(?<disallowed>[\w-~]+(?:\|[^:|()]+)*)\)\.\*$/.test(
          token.pattern.replace(/\\./g, "")
        )) {
          return `Path ${path7} cannot use unsupported regular expression wildcard. If the path includes special characters, they must be escaped with backslash (e.g. '\\(')`;
        }
        if (token.modifier && i !== tokens.length - 1) {
          return `Modifier ${token.modifier} is not allowed on wildcard :${token.name} in ${path7}. Modifiers are only allowed in the last path component`;
        }
      }
    }
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    return `Path ${path7} could not be parsed into regexp: ${message}`;
  }
  return void 0;
}
var validateAppPaths = (name, app) => {
  for (const group of app.routing) {
    for (const p of group.paths) {
      if (p === "/") {
        continue;
      }
      if (p.endsWith("/")) {
        throw new MicrofrontendError(
          `Invalid path for application "${name}". ${p} must not end with a slash.`,
          { type: "application", subtype: "invalid_path" }
        );
      }
      if (!p.startsWith("/")) {
        throw new MicrofrontendError(
          `Invalid path for application "${name}". ${p} must start with a slash.`,
          { type: "application", subtype: "invalid_path" }
        );
      }
    }
  }
  if (app.assetPrefix) {
    if (!VALID_ASSET_PREFIX_REGEXP.test(app.assetPrefix)) {
      throw new MicrofrontendError(
        `Invalid asset prefix for application "${name}". ${app.assetPrefix} must start with a lowercase letter and contain only lowercase letters, numbers, and hyphens.`,
        { type: "application", subtype: "invalid_asset_prefix" }
      );
    }
    if (app.assetPrefix !== `vc-ap-${name}` && !app.routing.some(
      (group) => group.paths.includes(`/${app.assetPrefix}/:path*`) && !group.flag
    )) {
      throw new MicrofrontendError(
        `When \`assetPrefix\` is specified, \`/${app.assetPrefix}/:path*\` must be added the routing paths for the application. Changing the asset prefix is not a forwards and backwards compatible change, and the custom asset prefix should be added to \`paths\` and deployed before setting the \`assetPrefix\` field.`,
        { type: "application", subtype: "invalid_asset_prefix" }
      );
    }
  }
};
var validateConfigDefaultApplication = (applicationConfigsById) => {
  if (!applicationConfigsById) {
    return;
  }
  const applicationsWithoutRouting = Object.entries(
    applicationConfigsById
  ).filter(([, app]) => isDefaultApp(app));
  const numApplicationsWithoutRouting = applicationsWithoutRouting.reduce(
    (acc) => {
      return acc + 1;
    },
    0
  );
  if (numApplicationsWithoutRouting === 0) {
    throw new MicrofrontendError(
      "No default application found. At least one application needs to be the default by omitting routing.",
      { type: "config", subtype: "no_default_application" }
    );
  }
  if (numApplicationsWithoutRouting > 1) {
    const applicationNamesMissingRouting = applicationsWithoutRouting.map(
      ([name]) => name
    );
    throw new MicrofrontendError(
      `All applications except for the default app must contain the "routing" field. Applications that are missing routing: ${LIST_FORMATTER.format(applicationNamesMissingRouting)}.`,
      { type: "config", subtype: "multiple_default_applications" }
    );
  }
};

// src/config/microfrontends-config/isomorphic/utils/hash-application-name.ts
var import_md5 = __toESM(require("md5"), 1);
function hashApplicationName(name) {
  if (!name) {
    throw new Error("Application name is required to generate hash");
  }
  return (0, import_md5.default)(name).substring(0, 6).padStart(6, "0");
}

// src/config/microfrontends-config/isomorphic/utils/generate-asset-prefix.ts
var PREFIX = "vc-ap";
function generateAssetPrefixFromName({
  name
}) {
  if (!name) {
    throw new Error("Name is required to generate an asset prefix");
  }
  return `${PREFIX}-${hashApplicationName(name)}`;
}

// src/config/microfrontends-config/isomorphic/utils/generate-port.ts
function generatePortFromName({
  name,
  minPort = 3e3,
  maxPort = 8e3
}) {
  if (!name) {
    throw new Error("Name is required to generate a port");
  }
  let hash = 0;
  for (let i = 0; i < name.length; i++) {
    hash = (hash << 5) - hash + name.charCodeAt(i);
    hash |= 0;
  }
  hash = Math.abs(hash);
  const range = maxPort - minPort;
  const port = minPort + hash % range;
  return port;
}

// src/config/microfrontends-config/isomorphic/host.ts
var Host = class {
  constructor(hostConfig, options) {
    if (typeof hostConfig === "string") {
      ({
        protocol: this.protocol,
        host: this.host,
        port: this.port
      } = Host.parseUrl(hostConfig));
    } else {
      const { protocol = "https", host, port } = hostConfig;
      this.protocol = protocol;
      this.host = host;
      this.port = port;
    }
    this.local = options?.isLocal;
  }
  static parseUrl(url, defaultProtocol = "https") {
    let hostToParse = url;
    if (!/^https?:\/\//.exec(hostToParse)) {
      hostToParse = `${defaultProtocol}://${hostToParse}`;
    }
    const parsed = new URL(hostToParse);
    if (!parsed.hostname) {
      throw new Error(Host.getMicrofrontendsError(url, "requires a host"));
    }
    if (parsed.hash) {
      throw new Error(
        Host.getMicrofrontendsError(url, "cannot have a fragment")
      );
    }
    if (parsed.username || parsed.password) {
      throw new Error(
        Host.getMicrofrontendsError(
          url,
          "cannot have authentication credentials (username and/or password)"
        )
      );
    }
    if (parsed.pathname !== "/") {
      throw new Error(Host.getMicrofrontendsError(url, "cannot have a path"));
    }
    if (parsed.search) {
      throw new Error(
        Host.getMicrofrontendsError(url, "cannot have query parameters")
      );
    }
    const protocol = parsed.protocol.slice(0, -1);
    return {
      protocol,
      host: parsed.hostname,
      port: parsed.port ? Number.parseInt(parsed.port) : void 0
    };
  }
  static getMicrofrontendsError(url, message) {
    return `Microfrontends configuration error: the URL ${url} in your microfrontends.json ${message}.`;
  }
  isLocal() {
    return this.local || this.host === "localhost" || this.host === "127.0.0.1";
  }
  toString() {
    const url = this.toUrl();
    return url.toString().replace(/\/$/, "");
  }
  toUrl() {
    const url = `${this.protocol}://${this.host}${this.port ? `:${this.port}` : ""}`;
    return new URL(url);
  }
};
var LocalHost = class extends Host {
  constructor({
    appName,
    local
  }) {
    let protocol;
    let host;
    let port;
    if (typeof local === "number") {
      port = local;
    } else if (typeof local === "string") {
      if (/^\d+$/.test(local)) {
        port = Number.parseInt(local);
      } else {
        const parsed = Host.parseUrl(local, "http");
        protocol = parsed.protocol;
        host = parsed.host;
        port = parsed.port;
      }
    } else if (local) {
      protocol = local.protocol;
      host = local.host;
      port = local.port;
    }
    super({
      protocol: protocol ?? "http",
      host: host ?? "localhost",
      port: port ?? generatePortFromName({ name: appName })
    });
  }
};

// src/config/microfrontends-config/isomorphic/utils/generate-automation-bypass-env-var-name.ts
function generateAutomationBypassEnvVarName({
  name
}) {
  return `AUTOMATION_BYPASS_${name.toUpperCase().replace(/[^a-zA-Z0-9]/g, "_")}`;
}

// src/config/microfrontends-config/isomorphic/application.ts
var Application = class {
  constructor(name, {
    app,
    overrides,
    isDefault
  }) {
    this.name = name;
    this.development = {
      local: new LocalHost({
        appName: name,
        local: app.development?.local
      }),
      fallback: app.development?.fallback ? new Host(app.development.fallback) : void 0
    };
    if (app.development?.fallback) {
      this.fallback = new Host(app.development.fallback);
    }
    this.packageName = app.packageName;
    this.overrides = overrides?.environment ? {
      environment: new Host(overrides.environment)
    } : void 0;
    this.default = isDefault ?? false;
    this.serialized = app;
  }
  isDefault() {
    return this.default;
  }
  getAssetPrefix() {
    const generatedAssetPrefix = generateAssetPrefixFromName({
      name: this.name
    });
    if ("assetPrefix" in this.serialized) {
      return this.serialized.assetPrefix ?? generatedAssetPrefix;
    }
    return generatedAssetPrefix;
  }
  getAutomationBypassEnvVarName() {
    return generateAutomationBypassEnvVarName({ name: this.name });
  }
  serialize() {
    return this.serialized;
  }
};
var DefaultApplication = class extends Application {
  constructor(name, {
    app,
    overrides
  }) {
    super(name, {
      app,
      overrides,
      isDefault: true
    });
    this.default = true;
    this.fallback = new Host(app.development.fallback);
  }
  getAssetPrefix() {
    return "";
  }
};
var ChildApplication = class extends Application {
  constructor(name, {
    app,
    overrides
  }) {
    ChildApplication.validate(name, app);
    super(name, {
      app,
      overrides,
      isDefault: false
    });
    this.default = false;
    this.routing = app.routing;
  }
  static validate(name, app) {
    validateAppPaths(name, app);
  }
};

// src/config/microfrontends-config/isomorphic/constants.ts
var DEFAULT_LOCAL_PROXY_PORT = 3024;

// src/config/microfrontends-config/isomorphic/index.ts
var MicrofrontendConfigIsomorphic = class {
  constructor({
    config,
    overrides
  }) {
    this.childApplications = {};
    MicrofrontendConfigIsomorphic.validate(config);
    const disableOverrides = config.options?.disableOverrides ?? false;
    this.overrides = overrides && !disableOverrides ? overrides : void 0;
    let defaultApplication;
    for (const [appId, appConfig] of Object.entries(config.applications)) {
      const appOverrides = !disableOverrides ? this.overrides?.applications[appId] : void 0;
      if (isDefaultApp(appConfig)) {
        defaultApplication = new DefaultApplication(appId, {
          app: appConfig,
          overrides: appOverrides
        });
      } else {
        this.childApplications[appId] = new ChildApplication(appId, {
          app: appConfig,
          overrides: appOverrides
        });
      }
    }
    if (!defaultApplication) {
      throw new MicrofrontendError(
        "Could not find default application in microfrontends configuration",
        {
          type: "application",
          subtype: "not_found"
        }
      );
    }
    this.defaultApplication = defaultApplication;
    this.config = config;
    this.options = config.options;
    this.serialized = {
      config,
      overrides
    };
  }
  static validate(config) {
    const c = typeof config === "string" ? (0, import_jsonc_parser.parse)(config) : config;
    validateConfigPaths(c.applications);
    validateConfigDefaultApplication(c.applications);
    return c;
  }
  static fromEnv({
    cookies
  }) {
    return new MicrofrontendConfigIsomorphic({
      config: (0, import_jsonc_parser.parse)(getConfigStringFromEnv()),
      overrides: parseOverrides(cookies ?? [])
    });
  }
  isOverridesDisabled() {
    return this.options?.disableOverrides ?? false;
  }
  getConfig() {
    return this.config;
  }
  getApplicationsByType() {
    return {
      defaultApplication: this.defaultApplication,
      applications: Object.values(this.childApplications)
    };
  }
  getChildApplications() {
    return Object.values(this.childApplications);
  }
  getAllApplications() {
    return [
      this.defaultApplication,
      ...Object.values(this.childApplications)
    ].filter(Boolean);
  }
  getApplication(name) {
    if (this.defaultApplication.name === name || this.defaultApplication.packageName === name) {
      return this.defaultApplication;
    }
    const app = this.childApplications[name] || Object.values(this.childApplications).find(
      (child) => child.packageName === name
    );
    if (!app) {
      throw new MicrofrontendError(
        `Could not find microfrontends configuration for application "${name}". If the name in package.json differs from your Vercel Project name, set the \`packageName\` field for the application in \`microfrontends.json\` to ensure that the configuration can be found locally.`,
        {
          type: "application",
          subtype: "not_found"
        }
      );
    }
    return app;
  }
  hasApplication(name) {
    try {
      this.getApplication(name);
      return true;
    } catch {
      return false;
    }
  }
  getApplicationByProjectName(projectName) {
    if (this.defaultApplication.name === projectName) {
      return this.defaultApplication;
    }
    return Object.values(this.childApplications).find(
      (app) => app.name === projectName
    );
  }
  /**
   * Returns the default application.
   */
  getDefaultApplication() {
    return this.defaultApplication;
  }
  /**
   * Returns the configured port for the local proxy
   */
  getLocalProxyPort() {
    return this.config.options?.localProxyPort ?? DEFAULT_LOCAL_PROXY_PORT;
  }
  toClientConfig(options) {
    const applications = Object.fromEntries(
      Object.entries(this.childApplications).map(([name, application]) => [
        hashApplicationName(name),
        {
          default: false,
          routing: application.routing
        }
      ])
    );
    applications[hashApplicationName(this.defaultApplication.name)] = {
      default: true
    };
    return new MicrofrontendConfigClient(
      {
        applications
      },
      {
        removeFlaggedPaths: options?.removeFlaggedPaths
      }
    );
  }
  /**
   * Serializes the class back to the Schema type.
   *
   * NOTE: This is used when writing the config to disk and must always match the input Schema
   */
  toSchemaJson() {
    return this.serialized.config;
  }
  serialize() {
    return this.serialized;
  }
};

// src/config/microfrontends/server/index.ts
var import_node_fs7 = __toESM(require("fs"), 1);
var import_node_path8 = require("path");

// src/config/microfrontends/utils/find-repository-root.ts
var import_node_fs = __toESM(require("fs"), 1);
var import_node_path = __toESM(require("path"), 1);
var GIT_DIRECTORY = ".git";
function hasGitDirectory(dir) {
  const gitPath = import_node_path.default.join(dir, GIT_DIRECTORY);
  return import_node_fs.default.existsSync(gitPath) && import_node_fs.default.statSync(gitPath).isDirectory();
}
function hasPnpmWorkspaces(dir) {
  return import_node_fs.default.existsSync(import_node_path.default.join(dir, "pnpm-workspace.yaml"));
}
function hasPackageJson(dir) {
  return import_node_fs.default.existsSync(import_node_path.default.join(dir, "package.json"));
}
function findRepositoryRoot(startDir) {
  if (process.env.NX_WORKSPACE_ROOT) {
    return process.env.NX_WORKSPACE_ROOT;
  }
  let currentDir = startDir || process.cwd();
  let lastPackageJsonDir = null;
  while (currentDir !== import_node_path.default.parse(currentDir).root) {
    if (hasGitDirectory(currentDir) || hasPnpmWorkspaces(currentDir)) {
      return currentDir;
    }
    if (hasPackageJson(currentDir)) {
      lastPackageJsonDir = currentDir;
    }
    currentDir = import_node_path.default.dirname(currentDir);
  }
  if (lastPackageJsonDir) {
    return lastPackageJsonDir;
  }
  throw new Error(
    `Could not find the root of the repository for ${startDir}. Please ensure that the directory is part of a Git repository. If you suspect that this should work, please file an issue to the Vercel team.`
  );
}

// src/config/microfrontends/utils/infer-microfrontends-location.ts
var import_node_path2 = require("path");
var import_node_fs2 = require("fs");
var import_jsonc_parser2 = require("jsonc-parser");
var import_fast_glob = __toESM(require("fast-glob"), 1);

// src/config/constants.ts
var CONFIGURATION_FILENAMES = [
  "microfrontends.jsonc",
  "microfrontends.json"
];

// src/config/microfrontends/utils/infer-microfrontends-location.ts
var configCache = {};
function findPackageWithMicrofrontendsConfig({
  repositoryRoot,
  applicationContext
}) {
  const applicationName = applicationContext.name;
  try {
    const microfrontendsJsonPaths = import_fast_glob.default.globSync(
      `**/{${CONFIGURATION_FILENAMES.join(",")}}`,
      {
        cwd: repositoryRoot,
        absolute: true,
        onlyFiles: true,
        followSymbolicLinks: false,
        ignore: ["**/node_modules/**", "**/.git/**"]
      }
    );
    const matchingPaths = [];
    for (const microfrontendsJsonPath of microfrontendsJsonPaths) {
      try {
        const microfrontendsJsonContent = (0, import_node_fs2.readFileSync)(
          microfrontendsJsonPath,
          "utf-8"
        );
        const microfrontendsJson = (0, import_jsonc_parser2.parse)(microfrontendsJsonContent);
        if (microfrontendsJson.applications[applicationName]) {
          matchingPaths.push(microfrontendsJsonPath);
        } else {
          for (const [_, app] of Object.entries(
            microfrontendsJson.applications
          )) {
            if (app.packageName === applicationName) {
              matchingPaths.push(microfrontendsJsonPath);
            }
          }
        }
      } catch (error) {
      }
    }
    if (matchingPaths.length > 1) {
      throw new MicrofrontendError(
        `Found multiple \`microfrontends.json\` files in the repository referencing the application "${applicationName}", but only one is allowed.
${matchingPaths.join("\n  \u2022 ")}`,
        { type: "config", subtype: "inference_failed" }
      );
    }
    if (matchingPaths.length === 0) {
      let additionalErrorMessage = "";
      if (microfrontendsJsonPaths.length > 0) {
        if (!applicationContext.projectName) {
          additionalErrorMessage = `

If the name in package.json (${applicationContext.packageJsonName}) differs from your Vercel Project name, set the \`packageName\` field for the application in \`microfrontends.json\` to ensure that the configuration can be found locally.`;
        } else {
          additionalErrorMessage = `

Names of applications in \`microfrontends.json\` must match the Vercel Project name (${applicationContext.projectName}).`;
        }
      }
      throw new MicrofrontendError(
        `Could not find a \`microfrontends.json\` file in the repository that contains the "${applicationName}" application.${additionalErrorMessage}

If your Vercel Microfrontends configuration is not in this repository, you can use the Vercel CLI to pull the Vercel Microfrontends configuration using the "vercel microfrontends pull" command, or you can specify the path manually using the VC_MICROFRONTENDS_CONFIG environment variable.

If you suspect this is thrown in error, please reach out to the Vercel team.`,
        { type: "config", subtype: "inference_failed" }
      );
    }
    const [packageJsonPath] = matchingPaths;
    return (0, import_node_path2.dirname)(packageJsonPath);
  } catch (error) {
    if (error instanceof MicrofrontendError) {
      throw error;
    }
    return null;
  }
}
function inferMicrofrontendsLocation(opts) {
  const cacheKey = `${opts.repositoryRoot}-${opts.applicationContext.name}`;
  if (configCache[cacheKey]) {
    return configCache[cacheKey];
  }
  const result = findPackageWithMicrofrontendsConfig(opts);
  if (!result) {
    throw new MicrofrontendError(
      `Could not infer the location of the \`microfrontends.json\` file for application "${opts.applicationContext.name}" starting in directory "${opts.repositoryRoot}".`,
      { type: "config", subtype: "inference_failed" }
    );
  }
  configCache[cacheKey] = result;
  return result;
}

// src/config/microfrontends/utils/is-monorepo.ts
var import_node_fs3 = __toESM(require("fs"), 1);
var import_node_path3 = __toESM(require("path"), 1);
function isMonorepo({
  repositoryRoot
}) {
  try {
    if (import_node_fs3.default.existsSync(import_node_path3.default.join(repositoryRoot, "pnpm-workspace.yaml"))) {
      return true;
    }
    if (import_node_fs3.default.existsSync(import_node_path3.default.join(repositoryRoot, "vlt-workspaces.json"))) {
      return true;
    }
    if (process.env.NX_WORKSPACE_ROOT === import_node_path3.default.resolve(repositoryRoot)) {
      return true;
    }
    const packageJsonPath = import_node_path3.default.join(repositoryRoot, "package.json");
    if (!import_node_fs3.default.existsSync(packageJsonPath)) {
      return false;
    }
    const packageJson = JSON.parse(
      import_node_fs3.default.readFileSync(packageJsonPath, "utf-8")
    );
    return packageJson.workspaces !== void 0;
  } catch (error) {
    console.error("Error determining if repository is a monorepo", error);
    return false;
  }
}

// src/config/microfrontends/utils/find-package-root.ts
var import_node_fs4 = __toESM(require("fs"), 1);
var import_node_path4 = __toESM(require("path"), 1);
var PACKAGE_JSON = "package.json";
function findPackageRoot(startDir) {
  let currentDir = startDir || process.cwd();
  while (currentDir !== import_node_path4.default.parse(currentDir).root) {
    const pkgJsonPath = import_node_path4.default.join(currentDir, PACKAGE_JSON);
    if (import_node_fs4.default.existsSync(pkgJsonPath)) {
      return currentDir;
    }
    currentDir = import_node_path4.default.dirname(currentDir);
  }
  throw new Error(
    `The root of the package that contains the \`package.json\` file for the \`${startDir}\` directory could not be found.`
  );
}

// src/config/microfrontends/utils/find-config.ts
var import_node_fs5 = __toESM(require("fs"), 1);
var import_node_path5 = require("path");
function findConfig({ dir }) {
  for (const filename of CONFIGURATION_FILENAMES) {
    const maybeConfig = (0, import_node_path5.join)(dir, filename);
    if (import_node_fs5.default.existsSync(maybeConfig)) {
      return maybeConfig;
    }
  }
  return null;
}

// src/config/microfrontends/utils/get-application-context.ts
var import_node_fs6 = __toESM(require("fs"), 1);
var import_node_path6 = __toESM(require("path"), 1);
function getApplicationContext(opts) {
  if (opts?.appName) {
    return { name: opts.appName };
  }
  if (process.env.VERCEL_PROJECT_NAME) {
    return {
      name: process.env.VERCEL_PROJECT_NAME,
      projectName: process.env.VERCEL_PROJECT_NAME
    };
  }
  if (process.env.NX_TASK_TARGET_PROJECT) {
    return {
      name: process.env.NX_TASK_TARGET_PROJECT,
      packageJsonName: process.env.NX_TASK_TARGET_PROJECT
    };
  }
  try {
    const vercelProjectJsonPath = import_node_fs6.default.readFileSync(
      import_node_path6.default.join(opts?.packageRoot || ".", ".vercel", "project.json"),
      "utf-8"
    );
    const projectJson = JSON.parse(vercelProjectJsonPath);
    if (projectJson.projectName) {
      return {
        name: projectJson.projectName,
        projectName: projectJson.projectName
      };
    }
  } catch (_) {
  }
  try {
    const packageJsonString = import_node_fs6.default.readFileSync(
      import_node_path6.default.join(opts?.packageRoot || ".", "package.json"),
      "utf-8"
    );
    const packageJson = JSON.parse(packageJsonString);
    if (!packageJson.name) {
      throw new MicrofrontendError(
        `package.json file missing required field "name"`,
        {
          type: "packageJson",
          subtype: "missing_field_name",
          source: "@vercel/microfrontends/next"
        }
      );
    }
    return { name: packageJson.name, packageJsonName: packageJson.name };
  } catch (err) {
    throw MicrofrontendError.handle(err, {
      fileName: "package.json"
    });
  }
}

// src/config/microfrontends/server/utils/get-output-file-path.ts
var import_node_path7 = __toESM(require("path"), 1);

// src/config/microfrontends/server/constants.ts
var MFE_CONFIG_DEFAULT_FILE_PATH = "microfrontends";
var MFE_CONFIG_DEFAULT_FILE_NAME = "microfrontends.json";

// src/config/microfrontends/server/utils/get-output-file-path.ts
function getOutputFilePath() {
  return import_node_path7.default.join(MFE_CONFIG_DEFAULT_FILE_PATH, MFE_CONFIG_DEFAULT_FILE_NAME);
}

// src/config/microfrontends/server/validation.ts
var import_jsonc_parser3 = require("jsonc-parser");
var import_ajv = require("ajv");

// schema/schema.json
var schema_default = {
  $schema: "http://json-schema.org/draft-07/schema#",
  $ref: "#/definitions/Config",
  definitions: {
    Config: {
      type: "object",
      properties: {
        $schema: {
          type: "string"
        },
        version: {
          type: "string",
          const: "1"
        },
        options: {
          $ref: "#/definitions/Options"
        },
        applications: {
          $ref: "#/definitions/ApplicationRouting",
          description: "Mapping of application names to the routes that they host. Only needs to be defined in the application that owns the primary microfrontend domain"
        }
      },
      required: [
        "applications"
      ],
      additionalProperties: false
    },
    Options: {
      type: "object",
      properties: {
        disableOverrides: {
          type: "boolean",
          description: "If you want to disable the overrides for the site. For example, if you are managing rewrites between applications externally, you may wish to disable the overrides on the toolbar as they will have no effect."
        },
        localProxyPort: {
          type: "number",
          description: "The port number used by the local proxy server.\n\nThe default is `3024`."
        }
      },
      additionalProperties: false
    },
    ApplicationRouting: {
      type: "object",
      additionalProperties: {
        $ref: "#/definitions/Application"
      },
      propertyNames: {
        description: "The unique identifier for a Microfrontend Application.\n\nMust match the Vercel project name.\n\nNote: If this name does not also match the name used to run the application, (e.g. the `name` from the `package.json`), then the `packageName` field should be set."
      }
    },
    Application: {
      anyOf: [
        {
          $ref: "#/definitions/DefaultApplication"
        },
        {
          $ref: "#/definitions/ChildApplication"
        }
      ]
    },
    DefaultApplication: {
      type: "object",
      properties: {
        packageName: {
          type: "string",
          description: "The name used to run the application, e.g. the `name` field in the `package.json`.\n\nThis is used by the local proxy to map the application config to the locally running app.\n\nThis is only necessary when the application name does not match the `name` used in `package.json`."
        },
        development: {
          $ref: "#/definitions/DefaultDevelopment",
          description: "Development configuration for the default application."
        }
      },
      required: [
        "development"
      ],
      additionalProperties: false
    },
    DefaultDevelopment: {
      type: "object",
      properties: {
        local: {
          type: [
            "number",
            "string"
          ],
          description: "A local port number or host string that this application runs on when it is running locally. If passing a string, include the protocol (optional), host (required) and port (optional). For example: `https://this.ismyhost:8080`. If omitted, the protocol defaults to HTTP. If omitted, the port defaults to a unique, but stable (based on the application name) number.\n\nExamples of valid values:\n- 8080\n- my.localhost.me\n- my.localhost.me:8080\n- https://my.localhost.me\n- https://my.localhost.me:8080"
        },
        task: {
          type: "string",
          description: "Optional task to run when starting the development server. Should reference a script in the package.json of the application."
        },
        fallback: {
          type: "string",
          description: "Fallback for local development, could point to any environment. This is required for the default app. This value is used as the fallback for child apps as well if they do not have a fallback.\n\nIf passing a string, include the protocol (optional), host (required) and port (optional). For example: `https://this.ismyhost:8080`. If omitted, the protocol defaults to HTTPS. If omitted, the port defaults to `80` for HTTP and `443` for HTTPS."
        }
      },
      required: [
        "fallback"
      ],
      additionalProperties: false
    },
    ChildApplication: {
      type: "object",
      properties: {
        packageName: {
          type: "string",
          description: "The name used to run the application, e.g. the `name` field in the `package.json`.\n\nThis is used by the local proxy to map the application config to the locally running app.\n\nThis is only necessary when the application name does not match the `name` used in `package.json`."
        },
        development: {
          $ref: "#/definitions/ChildDevelopment",
          description: "Development configuration for the child application."
        },
        routing: {
          $ref: "#/definitions/Routing",
          description: "Groups of path expressions that are routed to this application."
        },
        assetPrefix: {
          type: "string",
          description: "The name of the asset prefix to use instead of the auto-generated name.\n\nThe asset prefix is used to prefix all paths to static assets, such as JS, CSS, or images that are served by a specific application. It is necessary to ensure there are no conflicts with other applications on the same domain.\n\nAn auto-generated asset prefix of the form `vc-ap-<hash>` is used when this field is not provided.\n\nWhen this field is provided, `/${assetPrefix}/:path*` must also be added to the list of paths in the `routing` field. Changing the asset prefix after a microfrontend application has already been deployed is not a forwards and backwards compatible change, and the asset prefix should be added to the `routing` field and deployed before setting the `assetPrefix` field."
        }
      },
      required: [
        "routing"
      ],
      additionalProperties: false
    },
    ChildDevelopment: {
      type: "object",
      properties: {
        local: {
          type: [
            "number",
            "string"
          ],
          description: "A local port number or host string that this application runs on when it is running locally. If passing a string, include the protocol (optional), host (required) and port (optional). For example: `https://this.ismyhost:8080`. If omitted, the protocol defaults to HTTP. If omitted, the port defaults to a unique, but stable (based on the application name) number.\n\nExamples of valid values:\n- 8080\n- my.localhost.me\n- my.localhost.me:8080\n- https://my.localhost.me\n- https://my.localhost.me:8080"
        },
        task: {
          type: "string",
          description: "Optional task to run when starting the development server. Should reference a script in the package.json of the application."
        },
        fallback: {
          type: "string",
          description: "Fallback for local development, could point to any environment. This is optional for child apps. If not provided, the fallback of the default app will be used.\n\nIf passing a string, include the protocol (optional), host (required) and port (optional). For example: `https://this.ismyhost:8080`. If omitted, the protocol defaults to HTTPS. If omitted, the port defaults to `80` for HTTP and `443` for HTTPS."
        }
      },
      additionalProperties: false
    },
    Routing: {
      type: "array",
      items: {
        $ref: "#/definitions/PathGroup"
      }
    },
    PathGroup: {
      type: "object",
      properties: {
        group: {
          type: "string",
          description: "Optional group name for the paths"
        },
        flag: {
          type: "string",
          description: "flag name that can be used to enable/disable all paths in the group"
        },
        paths: {
          type: "array",
          items: {
            type: "string"
          }
        }
      },
      required: [
        "paths"
      ],
      additionalProperties: false
    }
  }
};

// src/config/schema/utils/load.ts
var SCHEMA = schema_default;

// src/config/microfrontends/server/validation.ts
var LIST_FORMATTER2 = new Intl.ListFormat("en", {
  style: "long",
  type: "disjunction"
});
function formatAjvErrors(errors) {
  if (!errors) {
    return [];
  }
  const errorMessages = [];
  for (const error of errors) {
    if (error.instancePath === "" && (error.keyword === "anyOf" || error.keyword === "required" && error.params.missingProperty === "partOf")) {
      continue;
    }
    const instancePath = error.instancePath.slice(1);
    const formattedInstancePath = instancePath === "" ? "at the root" : `in field ${instancePath}`;
    if (error.keyword === "required" && error.params.missingProperty === "routing" && instancePath.split("/").length === 2) {
      errorMessages.push(
        `Unable to infer if ${instancePath} is the default app or a child app. This usually means that there is another error in the configuration.`
      );
    } else if (error.keyword === "anyOf" && instancePath.split("/").length > 2) {
      const anyOfErrors = errors.filter(
        (e) => e.instancePath === error.instancePath && e.keyword !== "anyOf"
      );
      if (anyOfErrors.every((e) => e.keyword === "type")) {
        const allowedTypes = LIST_FORMATTER2.format(
          anyOfErrors.map((e) => {
            return e.keyword === "type" ? String(e.params.type) : "unknown";
          })
        );
        errorMessages.push(
          `Incorrect type for ${instancePath}. Must be one of ${allowedTypes}`
        );
      } else {
        errorMessages.push(
          `Invalid field for ${instancePath}. Possible error messages are ${LIST_FORMATTER2.format(anyOfErrors.map((e) => e.message ?? ""))}`
        );
      }
    } else if (error.keyword === "additionalProperties" && !(error.params.additionalProperty === "routing" && instancePath.split("/").length === 2)) {
      errorMessages.push(
        `Property '${error.params.additionalProperty}' is not allowed ${formattedInstancePath}`
      );
    } else if (error.keyword === "required") {
      errorMessages.push(
        `Property '${error.params.missingProperty}' is required ${formattedInstancePath}`
      );
    }
  }
  return errorMessages;
}
function validateSchema(configString) {
  const parsedConfig = (0, import_jsonc_parser3.parse)(configString);
  const ajv = new import_ajv.Ajv({ allowUnionTypes: true });
  const validate = ajv.compile(SCHEMA);
  const isValid = validate(parsedConfig);
  if (!isValid) {
    throw new MicrofrontendError(
      `Invalid microfrontends config:${formatAjvErrors(validate.errors).map((error) => `
 - ${error}`).join(
        ""
      )}

See https://openapi.vercel.sh/microfrontends.json for the schema.`,
      { type: "config", subtype: "does_not_match_schema" }
    );
  }
  return parsedConfig;
}

// src/config/microfrontends/server/index.ts
var MicrofrontendsServer = class {
  constructor({
    config,
    overrides
  }) {
    this.config = new MicrofrontendConfigIsomorphic({ config, overrides });
  }
  /**
   * Writes the configuration to a file.
   */
  writeConfig(opts = {
    pretty: true
  }) {
    const outputPath = getOutputFilePath();
    import_node_fs7.default.mkdirSync((0, import_node_path8.dirname)(outputPath), { recursive: true });
    import_node_fs7.default.writeFileSync(
      outputPath,
      JSON.stringify(
        this.config.toSchemaJson(),
        null,
        opts.pretty ?? true ? 2 : void 0
      )
    );
  }
  // --------- Static Methods ---------
  /**
   * Generates a MicrofrontendsServer instance from an unknown object.
   */
  static fromUnknown({
    config,
    cookies
  }) {
    const overrides = cookies ? parseOverrides(cookies) : void 0;
    if (typeof config === "string") {
      return new MicrofrontendsServer({
        config: MicrofrontendsServer.validate(config),
        overrides
      });
    }
    if (typeof config === "object") {
      return new MicrofrontendsServer({
        config,
        overrides
      });
    }
    throw new MicrofrontendError(
      "Invalid config: must be a string or an object",
      { type: "config", subtype: "does_not_match_schema" }
    );
  }
  /**
   * Generates a MicrofrontendsServer instance from the environment.
   * Uses additional validation that is only available when in a node runtime
   */
  static fromEnv({
    cookies
  }) {
    return new MicrofrontendsServer({
      config: MicrofrontendsServer.validate(getConfigStringFromEnv()),
      overrides: parseOverrides(cookies)
    });
  }
  /**
   * Validates the configuration against the JSON schema
   */
  static validate(config) {
    if (typeof config === "string") {
      const c = validateSchema(config);
      return c;
    }
    return config;
  }
  /**
   * Looks up the configuration by inferring the package root and looking for a microfrontends config file. If a file is not found,
   * it will look for a package in the repository with a microfrontends file that contains the current application
   * and use that configuration.
   *
   * This can return either a Child or Main configuration.
   */
  static infer({
    appName,
    directory,
    filePath,
    cookies
  } = {}) {
    if (filePath) {
      return MicrofrontendsServer.fromFile({
        filePath,
        cookies
      });
    }
    try {
      const packageRoot = findPackageRoot(directory);
      const applicationContext = getApplicationContext({
        appName,
        packageRoot
      });
      const maybeConfig = findConfig({ dir: packageRoot });
      if (maybeConfig) {
        return MicrofrontendsServer.fromFile({
          filePath: maybeConfig,
          cookies
        });
      }
      const repositoryRoot = findRepositoryRoot();
      const isMonorepo2 = isMonorepo({ repositoryRoot });
      if (typeof process.env.VC_MICROFRONTENDS_CONFIG === "string") {
        const maybeConfigFromEnv = (0, import_node_path8.resolve)(
          packageRoot,
          process.env.VC_MICROFRONTENDS_CONFIG
        );
        if (maybeConfigFromEnv) {
          return MicrofrontendsServer.fromFile({
            filePath: maybeConfigFromEnv,
            cookies
          });
        }
      } else {
        const maybeConfigFromVercel = findConfig({
          dir: (0, import_node_path8.join)(packageRoot, ".vercel")
        });
        if (maybeConfigFromVercel) {
          return MicrofrontendsServer.fromFile({
            filePath: maybeConfigFromVercel,
            cookies
          });
        }
        if (isMonorepo2) {
          const defaultPackage = inferMicrofrontendsLocation({
            repositoryRoot,
            applicationContext
          });
          const maybeConfigFromDefault = findConfig({ dir: defaultPackage });
          if (maybeConfigFromDefault) {
            return MicrofrontendsServer.fromFile({
              filePath: maybeConfigFromDefault,
              cookies
            });
          }
        }
      }
      throw new MicrofrontendError(
        'Unable to automatically infer the location of the `microfrontends.json` file. If your Vercel Microfrontends configuration is not in this repository, you can use the Vercel CLI to pull the Vercel Microfrontends configuration using the "vercel microfrontends pull" command, or you can specify the path manually using the VC_MICROFRONTENDS_CONFIG environment variable. If you suspect this is thrown in error, please reach out to the Vercel team.',
        { type: "config", subtype: "inference_failed" }
      );
    } catch (e) {
      if (e instanceof MicrofrontendError) {
        throw e;
      }
      const errorMessage = e instanceof Error ? e.message : String(e);
      throw new MicrofrontendError(
        `Unable to locate and parse the \`microfrontends.json\` configuration file. Original error message: ${errorMessage}`,
        { cause: e, type: "config", subtype: "inference_failed" }
      );
    }
  }
  /*
   * Generates a MicrofrontendsServer instance from a file.
   */
  static fromFile({
    filePath,
    cookies
  }) {
    try {
      const configJson = import_node_fs7.default.readFileSync(filePath, "utf-8");
      const config = MicrofrontendsServer.validate(configJson);
      return new MicrofrontendsServer({
        config,
        overrides: cookies ? parseOverrides(cookies) : void 0
      });
    } catch (e) {
      throw MicrofrontendError.handle(e, {
        fileName: filePath
      });
    }
  }
  /*
   * Generates a MicrofrontendsServer instance from a file.
   */
  static fromMainConfigFile({
    filePath,
    overrides
  }) {
    try {
      const config = import_node_fs7.default.readFileSync(filePath, "utf-8");
      const validatedConfig = MicrofrontendsServer.validate(config);
      const [defaultApplication] = Object.entries(validatedConfig.applications).filter(([, app]) => isDefaultApp(app)).map(([name]) => name);
      if (!defaultApplication) {
        throw new MicrofrontendError(
          "No default application found. At least one application needs to be the default by omitting routing.",
          { type: "config", subtype: "no_default_application" }
        );
      }
      return new MicrofrontendsServer({
        config: validatedConfig,
        overrides
      });
    } catch (e) {
      throw MicrofrontendError.handle(e, {
        fileName: filePath
      });
    }
  }
};

// src/bin/local-auth.ts
var localAuthHtml = ({
  app,
  hostname,
  defaultApp,
  automationBypassEnvVarName,
  automationBypass
}) => {
  const content = automationBypass ? `<b><code>${automationBypassEnvVarName}</code></b> is set with the value <b><code>${automationBypass}</code></b>, please verify this value equals the <a href="https://vercel.com/docs/deployment-protection/methods-to-bypass-deployment-protection/protection-bypass-automation">Protection Bypass for Automation</a> for the Vercel project hosting the deployment.` : `To access, set a local environment variable <b><code>${automationBypassEnvVarName}</code></b> with the value of the <a href="https://vercel.com/docs/deployment-protection/methods-to-bypass-deployment-protection/protection-bypass-automation">Protection Bypass for Automation</a> for the Vercel project hosting the deployment.`;
  return `<!DOCTYPE html>
	<html lang="en">
	  <head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1">
		<meta name="theme-color" content="#000">
		<title>Authentication Required</title>
		<style>
		  html {
			font-size: 62.5%;
			box-sizing: border-box;
			height: -webkit-fill-available
		  }
	
		  *,
		  ::after,
		  ::before {
			box-sizing: inherit
		  }
	
		  body {
			font-family: "SF Pro Text", "SF Pro Icons", "Helvetica Neue", "Helvetica", "Arial", sans-serif;
			font-size: 1.6rem;
			line-height: 1.65;
			word-break: break-word;
			font-kerning: auto;
			font-variant: normal;
			-webkit-font-smoothing: antialiased;
			-moz-osx-font-smoothing: grayscale;
			text-rendering: optimizeLegibility;
			hyphens: none;
			height: 100vh;
			height: -webkit-fill-available;
			max-height: 100vh;
			max-height: -webkit-fill-available;
			margin: 0
		  }
	
		  ::selection {
			background: #79FFE1;
		  }
		  ::-moz-selection {
			background: #79FFE1;
		  }
	
		  a {
			cursor: pointer;
			color: #0070f3;
			text-decoration: none;
			transition: all .2s ease;
			border-bottom: 1px solid transparent
		  }
	
		  a:hover {
			border-bottom: 1px solid #0070f3
		  }
	
		  ul {
			padding: 0;
			margin-left: 1.5em;
			list-style-type: none
		  }
	
		  li {
			margin-bottom: 10px
		  }
	
		  ul li:before {
			content: '\\02013'
		  }
	
		  li:before {
			display: inline-block;
			color: #ccc;
			position: absolute;
			margin-left: -18px;
			transition: color .2s ease
		  }
	
		  code {
			font-family: Menlo, Monaco, Lucida Console, Liberation Mono, DejaVu Sans Mono, Bitstream Vera Sans Mono, Courier New, monospace, serif;
			font-size: .92em;
		  }
	
		  code:after,
		  code:before {
			content: '\`'
		  }
	
		  .container {
			display: flex;
			justify-content: center;
			flex-direction: column;
			min-height: 100%
		  }
	
		  main {
			max-width: 80rem;
			padding: 4rem 6rem;
			margin: auto
		  }
	
		  ul {
			margin-bottom: 32px
		  }
	
		  .error-title {
			font-size: 2rem;
			padding-left: 22px;
			line-height: 1.5;
			margin-bottom: 24px;
		  }
	
		  .error-title-guilty {
			border-left: 2px solid #ED367F;
		  }
	
		  .error-title-innocent {
			border-left: 2px solid #59B89C;
		  }
	
		  main p {
			color: #333
		  }
	
		  .devinfo-container {
			border: 1px solid #ddd;
			border-radius: 4px;
			padding: 2rem;
			display: flex;
			flex-direction: column;
			margin-bottom: 32px
		  }
	
		  .error-code {
			margin: 0;
			font-size: 1.6rem;
			color: #000;
			margin-bottom: 1.6rem
		  }
	
		  .devinfo-line {
			color: #333
		  }
	
		  .devinfo-line code,
		  code,
		  li {
			color: #000
		  }
	
		  .devinfo-line:not(:last-child) {
			margin-bottom: 8px
		  }
	
		  .docs-link,
		  .contact-link {
			font-weight: 500
		  }
	
		  header,
		  footer,
		  footer a {
			display: flex;
			justify-content: center;
			align-items: center
		  }
	
		  header,
		  footer {
			min-height: 100px;
			height: 100px;
		  }
	
		  header {
			border-bottom: 1px solid #eaeaea;
		  }
	
		  header h1 {
			font-size: 1.8rem;
			margin: 0;
			font-weight: 500;
		  }
	
		  header p {
			font-size: 1.3rem;
			margin: 0;
			font-weight: 500;
		  }
	
		  .header-item {
			display: flex;
			padding: 0 2rem;
			margin: 2rem 0;
			text-decoration: line-through;
			color: #999;
		  }
	
		  .header-item.active {
			color: #ff0080;
			text-decoration: none;
		  }
	
		  .header-item.first {
			border-right: 1px solid #eaeaea;
		  }
	
		  .header-item-content {
			display: flex;
			flex-direction: column;
		  }
	
		  .header-item-icon {
			margin-right: 1rem;
			margin-top: 0.6rem;
		  }
	
		  footer {
			border-top: 1px solid #eaeaea;
		  }
	
		  footer a {
			color: #000
		  }
	
		  footer a:hover {
			border-bottom-color: transparent
		  }
	
		  footer svg {
			margin-left: .8rem
		  }
	
		  .note {
			padding: 8pt 16pt;
			border-radius: 5px;
			border: 1px solid #0070f3;
			font-size: 14px;
			line-height: 1.8;
			color: #0070f3;
		  }
	
		  @media (max-width:500px) {
			.devinfo-container .devinfo-line code {
			  margin-top: .4rem
			}
			.devinfo-container .devinfo-line:not(:last-child) {
			  margin-bottom: 1.6rem
			}
			.devinfo-container {
			  margin-bottom: 0;
			}
			header {
			  flex-direction: column;
			  height: auto;
			  min-height: auto;
			  align-items: flex-start;
			}
			.header-item.first {
			  border-right: none;
			  margin-bottom: 0;
			}
			main {
			  padding: 1rem 2rem;
			}
			body {
			  font-size: 1.4rem;
			  line-height: 1.55;
			}
			footer {
			  display: none;
			}
			.note {
			  margin-top: 16px;
			}
		  }
		</style>
	  </head>
	  <body>
		<div class="container">
		  <main>
			<p class="devinfo-container">
			  <span><b><code>${app}</code></b> is falling back to a protected deployment <a target="_blank" href="https://${hostname}">${hostname}</a>.</span>
        <br>
        <span>${content}</span>
        <br>
        <span>The environment variable should be set in the default app <b><code>${defaultApp}</code></b>, where the local proxy is running.</span>
			</p>
		
			<a href="https://vercel.com/docs/microfrontends/local-development#falling-back-to-protected-deployments"><div class="note">Click here to learn more about setting up the environment variable.</div></a>
		  </main>	
		</div>
	  </body>
	</html>`;
};

// src/bin/local-proxy.ts
var MFE_LOCAL_PROXY_HEADER = "x-vercel-mfe-local-proxy-origin";
var MFE_FLAG_VALUE = "vercel-mfe-flag-value";
var MFE_FLAG_VALUE_HEADER = `x-${MFE_FLAG_VALUE}`;
var ProxyRequestRouter = class {
  constructor(config, {
    localApps
  }) {
    this.config = config;
    this.localApps = localApps;
  }
  getDefaultHost(config) {
    const defaultApp = config.getDefaultApplication();
    return this.getApplicationTarget(defaultApp);
  }
  getApplicationTarget(application) {
    const useDev = Boolean(
      this.localApps.find(
        (name) => name === application.name || name === application.packageName
      )
    );
    let applicationName = application.name;
    let host = useDev ? application.development.local : application.fallback;
    if (application.overrides?.environment?.host) {
      host = application.overrides.environment;
    }
    if (!host) {
      const defaultApp = this.config.getDefaultApplication();
      host = defaultApp.fallback;
      applicationName = defaultApp.name;
    }
    const protocol = host.protocol;
    const hostname = host.host;
    const port = host.port;
    return {
      url: host.toUrl(),
      protocol,
      hostname,
      port,
      application: applicationName,
      isLocal: useDev,
      originalApplication: application.name
    };
  }
  /**
   * To enable preview deployments in localhost, we need to intercept some auth requests
   * and make sure they proxy to the correct domain. The toolbar will initiate the `vercel-auth-redirect`
   * with a `_host_override` param so that we can properly trigger the redirect flow for the
   * protected host.
   */
  getAuthTarget(request2, config) {
    const url = new import_node_url.URL(request2.url ?? "", `http://${request2.headers.host}`);
    const isAuthRedirect = request2.url?.startsWith(
      "/.well-known/vercel-auth-redirect"
    );
    const isSsoRedirect = request2.url?.startsWith("/sso-api");
    const isJWTRedirect = url.searchParams.has("_vercel_jwt");
    const defaultHost = this.getDefaultHost(config);
    let hostname = null;
    let path7 = request2.url;
    if (isAuthRedirect) {
      hostname = url.searchParams.get("_host_override");
    }
    if (isSsoRedirect) {
      hostname = "vercel.com";
    }
    if (isJWTRedirect) {
      hostname = url.searchParams.get("_host_override");
      url.searchParams.delete("_host_override");
      path7 = `${url.pathname}${url.search}`;
    }
    if (!hostname) {
      return void 0;
    }
    return { ...defaultHost, path: path7, hostname, protocol: "https", port: 443 };
  }
  getConfigWithOverrides(cookies) {
    const cookieOverrides = parseOverrides(
      Object.entries(cookies).map(([name, value]) => ({ name, value }))
    );
    const hasOverrides = Object.keys(cookieOverrides.applications).length > 0;
    const serialized = this.config.serialize().config;
    return hasOverrides ? new MicrofrontendConfigIsomorphic({
      config: serialized,
      overrides: cookieOverrides
    }) : this.config;
  }
  getTarget(request2) {
    const cookies = (0, import_cookie.parse)(request2.headers.cookie || "");
    const config = this.getConfigWithOverrides(cookies);
    if (!request2.url) {
      return this.getDefaultHost(config);
    }
    const { path: path7, mfeFlagValue } = extractMfeFlagValue(request2.url);
    const authTarget = this.getAuthTarget(request2, config);
    if (authTarget) {
      return authTarget;
    }
    const url = new import_node_url.URL(`http://example.com${path7}`);
    const target = this.findMatchingApplication({
      path: path7,
      url,
      applications: config.getChildApplications(),
      referer: request2.headers.referer,
      // If a request already has the local proxy header, then the request has
      // already gone through the local proxy.
      // This should only happen if middleware indicates that a flagged route
      // is enabled, so we treat this as enabling all flagged routes for this request.
      middlewareMfeZone: request2.headers["x-vercel-mfe-zone"],
      // Value to use when encountering any flagged paths instead of checking middleware
      mfeFlagValue
    });
    if (target)
      return target;
    const defaultHost = this.getDefaultHost(config);
    console.log(
      ` ${path7} - Did not match any routes. Routing to default app: ${formatProxyTarget(defaultHost)}`
    );
    return { path: path7, ...defaultHost };
  }
  findMatchingApplication({
    path: path7,
    url,
    applications,
    referer = void 0,
    middlewareMfeZone = void 0,
    mfeFlagValue = void 0
  }) {
    for (const application of Object.values(applications)) {
      const target = this.getApplicationTarget(application);
      if (middlewareMfeZone) {
        if (middlewareMfeZone === application.name) {
          console.log(
            ` ${path7} - Routing to ${formatProxyTarget(target)} according to 'x-vercel-mfe-zone' header`
          );
          return { path: path7, ...target };
        }
        continue;
      }
      const builtInRewrite = this.checkBuiltinAssetPrefix({
        rewrites: ["/:path*"],
        path: path7,
        url,
        app: application
      }) || this.checkNextOriginalFrame({ url, referer, applications }) || this.checkNextSourceMap({ url }) || this.checkNextImage({ url, applications });
      if (builtInRewrite) {
        return builtInRewrite;
      }
      for (const group of application.routing) {
        for (const childPath of group.paths) {
          const regexp = (0, import_path_to_regexp3.pathToRegexp)(childPath);
          if (regexp.test(url.pathname)) {
            if (group.flag) {
              if (mfeFlagValue === true) {
              } else if (mfeFlagValue === false) {
                continue;
              } else {
                console.log(
                  "Routing group is behind flag. Routing to default app to check flag via middleware."
                );
                if (!this.isDefaultAppLocal()) {
                  const defaultApp = this.getDefaultHost(this.config);
                  console.error(
                    `'${path7}' is a flagged path, but the default application is not running locally. Using '${defaultApp.hostname}' to handle this request.`
                  );
                }
                return null;
              }
            }
            console.log(
              ` ${path7} - Matched ${childPath}. Routing to ${formatProxyTarget(target)}`
            );
            return { path: path7, ...target };
          }
        }
      }
    }
    if (middlewareMfeZone) {
      console.error(
        `A request contained 'x-vercel-mfe-zone: ${middlewareMfeZone}', but no application was found with that name.`
      );
    }
    return null;
  }
  checkBuiltinAssetPrefix({
    rewrites,
    path: path7,
    url,
    app
  }) {
    const pathname = url.pathname;
    const target = this.getApplicationTarget(app);
    const assetPrefixes = /* @__PURE__ */ new Set([
      `vc-ap-${app.name}`,
      `vc-ap-${hashApplicationName(app.name)}`,
      app.getAssetPrefix()
    ]);
    for (const rewrite of rewrites) {
      for (const assetPrefix of assetPrefixes) {
        if ((0, import_path_to_regexp3.pathToRegexp)(`/${assetPrefix}${rewrite}`).test(pathname)) {
          console.log(
            ` ${pathname} - Matched asset prefix. Routing to ${formatProxyTarget(target)}`
          );
          return {
            path: path7,
            ...target
          };
        }
      }
    }
    return null;
  }
  checkNextOriginalFrame({
    url,
    referer = void 0,
    applications
  }) {
    const isStackFrame = (0, import_path_to_regexp3.pathToRegexp)("/__nextjs_original-stack-frame").test(url.pathname) || // Plural form was introduced in https://github.com/vercel/next.js/pull/75557
    (0, import_path_to_regexp3.pathToRegexp)("/__nextjs_original-stack-frames").test(url.pathname);
    if (!referer || !isStackFrame) {
      return null;
    }
    const refererURL = new import_node_url.URL(referer);
    const refererPath = `${refererURL.pathname}?${refererURL.search}`;
    const refererApp = this.findMatchingApplication({
      path: refererPath,
      url: refererURL,
      applications
    });
    if (!refererApp) {
      return null;
    }
    console.log(
      ` ${refererURL.pathname} - Routing nextjs stack frame request to ${formatProxyTarget(refererApp)}`
    );
    return {
      ...refererApp,
      path: `${url.pathname}${url.search}`
    };
  }
  checkNextSourceMap({ url }) {
    const isSourceMap = (0, import_path_to_regexp3.pathToRegexp)("/__nextjs_source-map").test(url.pathname);
    if (!isSourceMap) {
      return null;
    }
    const localApp = this.getArbitraryLocalApp();
    if (!localApp) {
      console.error(
        ` ${url.pathname} - No locally running application to route request to`
      );
      return null;
    }
    const target = this.getApplicationTarget(localApp);
    console.log(
      ` ${url.pathname} - Routing nextjs source map request to randomly selected local application: ${formatProxyTarget(target)}`
    );
    return {
      ...target,
      path: `${url.pathname}${url.search}`
    };
  }
  checkNextImage({
    url,
    applications
  }) {
    const isNextImage = (0, import_path_to_regexp3.pathToRegexp)("/_next/image").test(url.pathname);
    if (!isNextImage) {
      return null;
    }
    const imageUrl = url.searchParams.get("url");
    if (!imageUrl) {
      console.error(
        ` ${url.pathname}?${url.search} - No url parameter found in _next/image request`
      );
      return null;
    }
    const decodedPath = decodeURIComponent(imageUrl);
    const imageURL = new import_node_url.URL(`http://example.com${decodedPath}`);
    const imageApp = this.findMatchingApplication({
      path: decodedPath,
      url: imageURL,
      applications
    });
    if (!imageApp) {
      return null;
    }
    console.log(
      ` ${url.pathname}?${url.search} - Routing nextjs image request to ${formatProxyTarget(imageApp)}`
    );
    return {
      ...imageApp,
      path: `${url.pathname}${url.search}`
    };
  }
  isDefaultAppLocal() {
    const defaultApp = this.config.getDefaultApplication();
    return Boolean(
      this.localApps.find(
        (name) => name === defaultApp.name || name === defaultApp.packageName
      )
    );
  }
  getArbitraryLocalApp() {
    for (const application of this.config.getAllApplications()) {
      const name = application.name;
      if (this.localApps.includes(name)) {
        return application;
      }
    }
  }
};
var LocalProxy = class {
  constructor(config, {
    localApps,
    proxyPort,
    configFilePath
  }) {
    this.router = new ProxyRequestRouter(config, { localApps });
    this.proxyPort = proxyPort ?? this.router.config.getLocalProxyPort();
    this.configFilePath = configFilePath;
    this.proxy = import_http_proxy.default.createProxyServer({ secure: true });
    this.proxy.on("error", (err, req, res) => {
      if (res instanceof http.ServerResponse) {
        res.writeHead(500, {
          "Content-Type": "text/plain"
        });
      }
      const target = this.router.getTarget(req);
      res.end(
        `Error proxying request to ${formatProxyTarget(target)}. Is the server running locally on port ${target.port}?`
      );
      console.error(
        `Error proxying request for ${formatProxyTarget(target)}: `,
        err
      );
    });
  }
  static fromFile(filePath, {
    localApps,
    proxyPort
  }) {
    let microfrontends;
    if (filePath) {
      microfrontends = MicrofrontendsServer.fromFile({
        filePath
      });
    } else {
      microfrontends = MicrofrontendsServer.infer();
    }
    LocalProxy.validateLocalApps(localApps, microfrontends.config);
    return new LocalProxy(microfrontends.config, {
      localApps,
      proxyPort,
      configFilePath: filePath
    });
  }
  static validateLocalApps(localApps, config) {
    const unknownApps = [];
    const allApps = /* @__PURE__ */ new Set();
    for (const app of config.getAllApplications()) {
      allApps.add(app.name);
    }
    for (const app of localApps) {
      if (!allApps.has(app)) {
        unknownApps.push(app);
      }
    }
    if (unknownApps.length) {
      throw new Error(
        `The following apps passed via --local-apps are not in the microfrontends config: ${unknownApps.join(", ")} (microfrontends config contains: ${Array.from(allApps).join(", ")})`
      );
    }
  }
  startServer() {
    const httpServer = http.createServer(
      (req, res) => this.handleRequest(req, res)
    );
    httpServer.on("upgrade", (req, socket, head) => {
      const target = this.router.getTarget(req);
      try {
        const headers = {};
        headers[MFE_LOCAL_PROXY_HEADER] = "1";
        this.proxy.ws(req, socket, head, {
          target: target.url,
          headers
        });
      } catch (err) {
        console.error("Error proxying ws: ", err);
      }
    });
    httpServer.listen(this.proxyPort, () => {
      this.displayStartupMessage();
    });
  }
  handleRequest(req, res) {
    if (this.handleProxyInfoRequest(req.url, res)) {
      return;
    }
    if (req.url?.includes("//")) {
      const originalUrl = req.url;
      if (originalUrl) {
        const normalizedUrl = originalUrl.replaceAll(/\/[\\/]+/g, "/");
        if (normalizedUrl !== originalUrl) {
          res.writeHead(307, {
            Location: normalizedUrl,
            // Copy incoming request headers except hop-by-hop headers and Location
            ...Object.fromEntries(
              Object.entries(req.headers).filter(
                ([key]) => ![
                  "connection",
                  "content-length",
                  "transfer-encoding",
                  "location"
                ].includes(key.toLowerCase())
              )
            )
          });
          res.end();
          return;
        }
      }
    }
    const target = this.router.getTarget(req);
    const { req: strippedReq, mfeFlagValue } = removeMfeFlagQuery(req);
    if (target.protocol === "https") {
      const { hostname, port, path: path7 } = target;
      const app = this.router.config.getApplication(target.application);
      const automationBypass = process.env[app.getAutomationBypassEnvVarName()];
      const requestOptions = {
        hostname,
        path: path7,
        method: req.method,
        headers: {
          ...req.headers,
          host: hostname,
          ...automationBypass ? { "x-vercel-protection-bypass": automationBypass } : {}
        },
        port
      };
      const localhost = `http://localhost:${this.proxyPort}`;
      const proxyReq = https.request(requestOptions, (realRes) => {
        if (realRes.statusCode === 401 && realRes.headers["set-cookie"]?.find(
          (cookie) => cookie.startsWith("_vercel_sso_nonce=")
        )) {
          const defaultApp = this.router.config.getDefaultApplication();
          return res.end(
            localAuthHtml({
              app: target.application,
              hostname,
              defaultApp: defaultApp.packageName || defaultApp.name,
              automationBypassEnvVarName: app.getAutomationBypassEnvVarName(),
              automationBypass
            })
          );
        }
        if (realRes.statusCode === 307 || realRes.statusCode === 308 || realRes.statusCode === 302 || realRes.statusCode === 301) {
          const locationHeader = realRes.headers.location;
          if (locationHeader) {
            const redirectUrl = new import_node_url.URL(
              locationHeader.replace(/https:\/\/[^/]+\//, "/"),
              localhost
            );
            realRes.headers.location = redirectUrl.toString();
          }
        }
        res.writeHead(realRes.statusCode || 200, realRes.headers);
        realRes.pipe(res, { end: true });
      });
      req.pipe(proxyReq);
      proxyReq.on("error", (err) => {
        console.error("Proxy request error: ", err);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end(
          `Error proxying request for ${target.application} to ${hostname}:${port}${path7}`
        );
      });
    } else {
      const headers = {};
      headers[MFE_LOCAL_PROXY_HEADER] = "1";
      if (mfeFlagValue !== void 0) {
        headers[MFE_FLAG_VALUE_HEADER] = mfeFlagValue.toString();
      }
      this.proxy.web(strippedReq, res, {
        target: target.url,
        headers
      });
    }
  }
  // Handles requests that return data from the local proxy itself.
  // Returns true if the request was handled, false otherwise.
  handleProxyInfoRequest(path7, res) {
    if (!path7) {
      return false;
    }
    const url = new import_node_url.URL(`http://example.comf${path7}`);
    const pathname = url.pathname;
    switch (pathname) {
      case "/.well-known/vercel/microfrontends/routing": {
        res.writeHead(200, {
          "Content-Type": "application/json"
        });
        const payload = Object.fromEntries(
          this.router.config.getAllApplications().map((app) => {
            const { hostname, port, protocol } = this.router.getApplicationTarget(app);
            return [
              app.name,
              { routing: { host: hostname, port, protocol } }
            ];
          })
        );
        res.end(JSON.stringify(payload));
        return true;
      }
    }
    return false;
  }
  displayStartupMessage() {
    const allApps = this.router.config.getAllApplications();
    const localApps = [];
    const fallbackApps = [];
    const defaultApp = this.router.config.getDefaultApplication();
    const defaultFallback = defaultApp.fallback.host;
    for (const app of allApps) {
      const isLocal = this.router.localApps.find(
        (name) => name === app.name || name === app.packageName
      );
      if (isLocal) {
        localApps.push({
          name: app.name,
          port: app.development.local.port
        });
      } else {
        const target = this.router.getApplicationTarget(app);
        let fallbackHost = target.hostname;
        if (!app.fallback) {
          fallbackHost = defaultFallback;
        }
        fallbackApps.push({
          name: app.name,
          fallback: fallbackHost
        });
      }
    }
    console.log(`
\u25B2 Microfrontends Proxy (${package_default.version}) Started`);
    console.log(`  - Proxy URL: http://localhost:${this.proxyPort}`);
    if (this.configFilePath) {
      console.log(`  - Config: ${this.configFilePath}`);
    }
    if (localApps.length > 0) {
      console.log("  - Local Applications:");
      const displayLocalApps = localApps.length > 5 ? [
        ...localApps.slice(0, 5),
        { name: `... and ${localApps.length - 5} more`, port: void 0 }
      ] : localApps;
      for (const app of displayLocalApps) {
        if (app.port !== void 0) {
          console.log(`    \u2022 ${app.name} (port ${app.port})`);
        } else {
          console.log(`    \u2022 ${app.name}`);
        }
      }
    }
    if (fallbackApps.length > 0) {
      console.log("  - Fallback Applications:");
      const displayFallbackApps = fallbackApps.length > 5 ? [
        ...fallbackApps.slice(0, 5),
        { name: `... and ${fallbackApps.length - 5} more`, fallback: "" }
      ] : fallbackApps;
      for (const app of displayFallbackApps) {
        if (app.fallback) {
          console.log(`    \u2022 ${app.name} \u2192 ${app.fallback}`);
        } else {
          console.log(`    \u2022 ${app.name}`);
        }
      }
    }
    if (localApps.length === 0 && fallbackApps.length === 0) {
      console.log("  - No applications configured\n");
    }
    if (localApps.length > 0) {
      console.log(
        "\nRequests directly to the ports of these applications may be automatically\nredirected to this proxy. Set the MFE_DISABLE_LOCAL_PROXY_REWRITE=1\nenvironment variable to disable this behavior."
      );
    }
    console.log(`
${"\u2500".repeat(50)}
`);
  }
};
function extractMfeFlagValue(path7) {
  const host = "http://example.com";
  const url = new import_node_url.URL(`${host}${path7}`);
  const mfeFlagValue = stringAsBoolean(url.searchParams.get(MFE_FLAG_VALUE));
  url.searchParams.delete(MFE_FLAG_VALUE);
  const pathWithoutMfe = mfeFlagValue === void 0 ? path7 : url.toString().substring(host.length);
  return { path: pathWithoutMfe, mfeFlagValue };
}
function stringAsBoolean(value) {
  if (value === "true") {
    return true;
  }
  if (value === "false") {
    return false;
  }
  return void 0;
}
function removeMfeFlagQuery(req) {
  if (!req.url) {
    return { req };
  }
  const { path: path7, mfeFlagValue } = extractMfeFlagValue(req.url);
  if (mfeFlagValue === void 0) {
    return { req };
  }
  req.url = path7;
  return { req, mfeFlagValue };
}
function formatProxyTarget(target) {
  return `${target.originalApplication} (${target.isLocal ? "local" : "fallback"})`;
}

// src/bin/port.ts
var import_node_process = require("process");

// src/utils/mfe-port.ts
var import_node_path9 = __toESM(require("path"), 1);
var import_node_fs8 = __toESM(require("fs"), 1);
function mfePort(packageDir) {
  const { name: appName, version } = getPackageJson(packageDir);
  try {
    const result = loadConfig({ packageDir, appName });
    const { port } = result;
    return {
      name: appName,
      version,
      port
    };
  } catch (e) {
    throw new Error(`Unable to determine configured port for ${appName}`, {
      cause: e
    });
  }
}
function getPackageJson(packageDir) {
  const filePath = import_node_path9.default.join(packageDir, "package.json");
  return JSON.parse(import_node_fs8.default.readFileSync(filePath, "utf-8"));
}
function loadConfig({
  packageDir,
  appName
}) {
  const config = MicrofrontendsServer.infer({
    directory: packageDir
  });
  const app = config.config.getApplication(appName);
  const port = app.development.local.port ?? (app.development.local.protocol === "https" ? 443 : 80);
  return { port };
}

// src/bin/port.ts
function displayPort() {
  const portInfo = mfePort((0, import_node_process.cwd)());
  header(portInfo);
  console.log(portInfo.port);
}
function header({ name, version, port }) {
  console.error(`
  \u25B2 ${name}@${version}
    \xB7 setting port to ${port}
  `);
}

// src/bin/index.ts
function main() {
  (0, import_env.loadEnvConfig)(process.cwd());
  const program = new import_commander.Command();
  program.name(package_default.name).description("Tools for working with micro-frontend applications").version(package_default.version, "-v, --version", "output the current version");
  program.command("proxy").argument("[filePath]", "Path to the micro-frontend configuration file").option(
    "--local-apps <names...>",
    "List of locally running applications",
    []
  ).addOption(new import_commander.Option("--names <names...>").hideHelp()).option("--port <port>", "Port proxy will use", (value) => {
    const parsedValue = Number.parseInt(value, 10);
    if (isNaN(parsedValue) || parsedValue <= 0) {
      throw new Error(
        "The value passed in to --port must be a positive number."
      );
    }
    return parsedValue;
  }).action((filePath, options) => {
    if (options.names && options.localApps.length) {
      throw new Error(
        "Both --names and --local-apps are set. --names is deprecated and has been replaced with --local-apps, which functions exactly the same. Please only set --local-apps."
      );
    }
    const localProxy = LocalProxy.fromFile(filePath, {
      localApps: options.names ?? options.localApps,
      proxyPort: options.port
    });
    localProxy.startServer();
  });
  program.command("port").description("Prints development port").action(() => {
    displayPort();
  });
  program.parse(process.argv);
}
main();
